#
# Submitted:
#
# Marc Kleine-Budde <kleine-budde@gmx.de>, 2004/05/03
# actually this patch is stolen from famous Dan Kegel's crosstool
#
# Error:
#
# glibc-2.3.2 with gcc > 3.4
#
# i386-unknown-linux-gnu-gcc dl-runtime.c -c -std=gnu99 -O2 -Wall
# -Winline -Wstrict-prototypes -Wwrite-strings -g -I../include
# -I. -I/home/frogger/ng-ptxdist/ptxdist-i386-generic-glibc/build/glibc-2.3.2-build/elf
# -I.. -I../libio
# -I/home/frogger/ng-ptxdist/ptxdist-i386-generic-glibc/build/glibc-2.3.2-build
# -I../sysdeps/i386/elf -I../linuxthreads/sysdeps/unix/sysv/linux/i386
# -I../linuxthreads/sysdeps/unix/sysv/linux
# -I../linuxthreads/sysdeps/pthread -I../sysdeps/pthread
# -I../linuxthreads/sysdeps/unix/sysv -I../linuxthreads/sysdeps/unix
# -I../linuxthreads/sysdeps/i386 -I../sysdeps/unix/sysv/linux/i386
# -I../sysdeps/unix/sysv/linux -I../sysdeps/gnu
# -I../sysdeps/unix/common -I../sysdeps/unix/mman
# -I../sysdeps/unix/inet -I../sysdeps/unix/sysv/i386
# -I../sysdeps/unix/sysv -I../sysdeps/unix/i386 -I../sysdeps/unix
# -I../sysdeps/posix -I../sysdeps/i386/fpu -I../sysdeps/i386
# -I../sysdeps/wordsize-32 -I../sysdeps/ieee754/ldbl-96
# -I../sysdeps/ieee754/dbl-64 -I../sysdeps/ieee754/flt-32
# -I../sysdeps/ieee754 -I../sysdeps/generic/elf -I../sysdeps/generic
# -nostdinc -isystem
# /home/frogger/projects/ng-ptxdist/xchain/i386-generic-glibc/bin/../lib/gcc/i386-unknown-linux-gnu/3.4.0/include
# -isystem
# /home/frogger/ng-ptxdist/xchain/i386-generic-glibc/i386-unknown-linux-gnu/include
# -D_LIBC_REENTRANT -include ../include/libc-symbols.h -o
# /home/frogger/ng-ptxdist/ptxdist-i386-generic-glibc/build/glibc-2.3.2-build/elf/dl-runtime.o
# dl-runtime.c:56: error: conflicting types for 'fixup'
# ../sysdeps/i386/dl-machine.h:158: error: previous declaration of
# 'fixup' was here dl-runtime.c:56: error: conflicting types for
# 'fixup' ../sysdeps/i386/dl-machine.h:158: error: previous
# declaration of 'fixup' was here dl-runtime.c:141: error: conflicting
# types for 'profile_fixup' ../sysdeps/i386/dl-machine.h:161: error:
# previous declaration of 'profile_fixup' was here dl-runtime.c:141:
# error: conflicting types for 'profile_fixup'
# ../sysdeps/i386/dl-machine.h:161: error: previous declaration of
# 'profile_fixup' was here ../sysdeps/i386/dl-machine.h:158: warning:
# 'fixup' declared `static' but never defined
# ../sysdeps/i386/dl-machine.h:161: warning: 'profile_fixup' declared
# `static' but never defined make[3]: ***
# [/home/frogger/ng-ptxdist/ptxdist-i386-generic-glibc/build/glibc-2.3.2-build/elf/dl-runtime.o]
# Error 1
#
# State:
#
# fixed in CVS
#
# Comments
#
# First hunk:
# Define ARCH_FIXUP_ATTRIBUTE and use it in the fixup function declarations.
# http://sources.redhat.com/cgi-bin/cvsweb.cgi/libc/sysdeps/i386/dl-machine.h.diff?r1=1.124&r2=1.125&cvsroot=glibc
# [rediffed against glibc-2.2.5]
#
# Second hunk:
# If ARCH_FIXUP_ATTRIBUTE is not defined, provide dummy definition.
# Use macro in fixup function definitions.
# http://sources.redhat.com/cgi-bin/cvsweb.cgi/libc/elf/dl-runtime.c.diff?r1=1.64&r2=1.65&cvsroot=glibc
# [rediffed against glibc-2.3.2]
#
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/i386/dl-machine.h,v
retrieving revision 1.124
retrieving revision 1.125
diff -u -r1.124 -r1.125
--- libc/sysdeps/i386/dl-machine.h	2004/03/05 10:14:49	1.124
+++ libc/sysdeps/i386/dl-machine.h	2004/03/09 07:42:29	1.125
@@ -154,11 +154,14 @@
    destroys the passed register information.  */
 /* GKM FIXME: Fix trampoline to pass bounds so we can do
    without the `__unbounded' qualifier.  */
-static ElfW(Addr) fixup (struct link_map *__unbounded l, ElfW(Word) reloc_offset)
-     __attribute__ ((regparm (2), unused));
+#define ARCH_FIXUP_ATTRIBUTE __attribute__ ((regparm (3), unused))
+
+static ElfW(Addr) fixup (struct link_map *__unbounded l,
+			 ElfW(Word) reloc_offset)
+     ARCH_FIXUP_ATTRIBUTE;
 static ElfW(Addr) profile_fixup (struct link_map *l, ElfW(Word) reloc_offset,
 				 ElfW(Addr) retaddr)
-     __attribute__ ((regparm (3), unused));
+     ARCH_FIXUP_ATTRIBUTE;
 # endif
 
 /* This code is used in dl-runtime.c to call the `fixup' function
===================================================================
--- /home/dank/downloads/glibc-2.3.2/elf/dl-runtime.c	Fri Feb  7 11:41:12 2003
+++ glibc-2.3.2/elf/dl-runtime.c	Thu Apr  8 22:24:26 2004
@@ -36,6 +36,12 @@
 # define VERSYMIDX(sym)	(DT_NUM + DT_THISPROCNUM + DT_VERSIONTAGIDX (sym))
 #endif
 
+/* The fixup functions might have need special attributes.  If none
+   are provided define the macro as empty.  */
+#ifndef ARCH_FIXUP_ATTRIBUTE
+# define ARCH_FIXUP_ATTRIBUTE
+#endif
+
 
 /* This function is called through a special trampoline from the PLT the
    first time each PLT entry is called.  We must perform the relocation
@@ -45,7 +51,7 @@
    function.  */
 
 #ifndef ELF_MACHINE_NO_PLT
-static ElfW(Addr) __attribute_used__
+static ElfW(Addr) __attribute_used__ ARCH_FIXUP_ATTRIBUTE
 fixup (
 # ifdef ELF_MACHINE_RUNTIME_FIXUP_ARGS
         ELF_MACHINE_RUNTIME_FIXUP_ARGS,
@@ -132,7 +138,7 @@
 
 #if !defined PROF && !defined ELF_MACHINE_NO_PLT && !__BOUNDED_POINTERS__
 
-static ElfW(Addr) __attribute_used__
+static ElfW(Addr) __attribute_used__ ARCH_FIXUP_ATTRIBUTE
 profile_fixup (
 #ifdef ELF_MACHINE_RUNTIME_FIXUP_ARGS
        ELF_MACHINE_RUNTIME_FIXUP_ARGS,
